src/attributes.rs

```diff
@@ -0,0 +1,282 @@
+use std::rc::Rc;
+
+#[derive(Debug, Clone, PartialEq, Eq)]
+pub struct Attribute<'a> {
+    pub is_inner: bool,
+    pub content: Rc<AttributeMetaItem<'a>>,
+}
+
+impl<'a> Attribute<'a> {
+    pub fn as_string(&self) -> String {
+        format!(
+            "#{}[{}]",
+            if self.is_inner { "!" } else { "" },
+            self.content.raw_item
+        )
+    }
+
+    pub fn new(raw: &'a str) -> Self {
+        let raw_trimmed = raw.trim();
+        let raw_without_closing = raw_trimmed.strip_suffix(']').unwrap_or_else(|| {
+            panic!(
+                concat!(
+                    "String `{}` cannot be parsed as an attribute ",
+                    "because it is not closed with a square bracket."
+                ),
+                raw_trimmed
+            )
+        });
+
+        if let Some(raw_content) = raw_without_closing.strip_prefix("#[") {
+            Attribute {
+                is_inner: false,
+                content: Rc::new(AttributeMetaItem::new(raw_content)),
+            }
+        } else if let Some(raw_content) = raw_without_closing.strip_prefix("#![") {
+            Attribute {
+                is_inner: true,
+                content: Rc::new(AttributeMetaItem::new(raw_content)),
+            }
+        } else {
+            panic!(
+                concat!(
+                    "String `{}` cannot be parsed as an attribute ",
+                    "because it starts with neither `#[` nor `#![`."
+                ),
+                raw_trimmed
+            )
+        }
+    }
+}
+
+#[derive(Debug, Clone, PartialEq, Eq)]
+pub struct AttributeMetaItem<'a> {
+    pub raw_item: &'a str,
+    pub base: &'a str,
+    pub assigned_item: Option<&'a str>,
+    pub arguments: Option<Vec<Rc<AttributeMetaItem<'a>>>>,
+}
+
+impl<'a> AttributeMetaItem<'a> {
+    /// Tries to parse `raw` as a comma-separated sequence of `AttributeMetaItem`'s
+    /// wrapped in parentheses, square brackets or curly brackets.
+    fn slice_arguments(raw: &'a str) -> Option<Vec<Rc<AttributeMetaItem<'a>>>> {
+        let is_left = |c| c == '(' || c == '[' || c == '{';
+        let is_right = |c| c == ')' || c == ']' || c == '}';
+        let matching_right = |c| match c {
+            '(' => ')',
+            '[' => ']',
+            '{' => '}',
+            _ => unreachable!("Tried to find matching right bracket for {}.", c),
+        };
+
+        let raw_trimmed = raw.trim();
+        let first_char = raw_trimmed.chars().next()?;
+        let raw_meta_seq = raw_trimmed
+            .strip_prefix(is_left)?
+            .strip_suffix(|c| c == matching_right(first_char))?
+            .trim();
+
+        let mut i = 0; // index of the first character after the last comma
+        let mut bracket_depth = 0; // number of currently opened brackets
```

Comment:
Consider keeping a stack (`Vec<char>`, since pushing and popping off the end are cheap) of brackets instead of just the depth, so you can make sure that you are tracking the depth correctly and closing the correct bracket scopes, or panic if an unexpected bracket is encountered. Then `bracket_depth` is just the length of that stack.

Reply:
Is encountering an unexpected bracket a good reason to panic? Or maybe it would be better to just return the `AttributeMetaItem` with `base=raw_item` and `assigned_item=arguments=None` as we discussed earlier? It is still possible that there is some rare syntax of attributes that is not covered by this schema.

Reply:
This is a tricky question, and in the end comes down to a judgment call.

Ideally, we don't want either panics nor incompletely-parsed data with `None` values. We'll want to proactively screen crates for these unusual pieces of syntax, and Tomasz is working on a way to run `cargo-semver-checks` on all the top crates on crates.io.

A panic would make it easier to spot one of those strange pieces of syntax, since the program would crash. But if we don't panic, we could still write a query that looks for e.g. `[` or `(` or `{` characters inside `base` and find those instances as well. And not panicking means we won't immediately blow up downstream code: for example, `cargo-semver-checks` can continue checking with all the other lints, and may merely suffer some false-negatives where attributes that couldn't be parsed were hiding semver issues. I don't see another argument in favor of panicking instead.

So I'd argue in favor of not panicking.

Reply:
I definitely agree with you. The only case for panicking now is when the type of attribute cannot be recognized, which I think is reasonable. 

I changed `bracket_depth` to `Vec<char>` as you suggested.